blobstore/data/file_repo.go

```diff
@@ -0,0 +1,434 @@
+		if !strings.HasPrefix(env, configEnvPrefix) {
+			continue
+		}
+		nameInfo := strings.SplitN(env, "=", 2)
+		name := strings.ToLower(strings.TrimPrefix(nameInfo[0], configEnvPrefix))
+		info := parseBackendInfo(nameInfo[1])
+		if info["backend"] != "s3" {
+			return nil, fmt.Errorf("blobstore: unknown backend %q for %s", info["backend"], name)
+		}
+		b, err := backend.NewS3(name, info)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Configured additional backend: %s (%s)\n", name, info["backend"])
+		backends = append(backends, b)
+	}
+
+	defaultBackend := "postgres"
+	if d := os.Getenv("DEFAULT_BACKEND"); d != "" {
+		defaultBackend = d
+		var found bool
+		for _, b := range backends {
+			if b.Name() == d {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("error: unknown default backend %q", d)
+		}
+	}
+	return NewFileRepo(db, backends, defaultBackend), nil
+}
+
+func parseBackendInfo(s string) map[string]string {
+	info := make(map[string]string)
+	for _, token := range strings.Split(s, " ") {
+		kv := strings.SplitN(token, "=", 2)
+		info[kv[0]] = kv[1]
+	}
+	return info
+}
+
+type FileRepo struct {
+	db             *postgres.DB
+	backends       map[string]backend.Backend
+	defaultBackend backend.Backend
+}
+
+func NewFileRepo(db *postgres.DB, backends []backend.Backend, defaultBackend string) *FileRepo {
+	r := &FileRepo{
+		db:       db,
+		backends: make(map[string]backend.Backend, len(backends)),
+	}
+	for _, b := range backends {
+		r.backends[b.Name()] = b
+	}
+	r.defaultBackend = r.backends[defaultBackend]
+	return r
+}
+
+func (r *FileRepo) getBackend(name string) (backend.Backend, error) {
+	if b, ok := r.backends[name]; ok {
+		return b, nil
+	}
+	return nil, fmt.Errorf("blobstore: unknown backend %q", name)
+}
+
+func (r *FileRepo) DefaultBackend() backend.Backend {
+	return r.defaultBackend
+}
+
+func (r *FileRepo) List(dir string) ([]string, error) {
+	rows, err := r.db.Query("SELECT substring(name FROM '^' || $1 || '/[^/]+/?') AS path FROM files WHERE name LIKE $1 || '%' AND deleted_at IS NULL GROUP BY path", strings.TrimSuffix(dir, "/"))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var paths []string
+	for rows.Next() {
+		var path *string
+		if err := rows.Scan(&path); err != nil {
+			return nil, err
+		}
+		if path != nil {
+			paths = append(paths, *path)
+		}
+	}
+	return paths, rows.Err()
+}
+
+type BackendFile struct {
+	backend.Backend
+	backend.FileInfo
+}
+
+func (r *FileRepo) ListFilesExcludingDefaultBackend(prefix string) ([]BackendFile, error) {
+	rows, err := r.db.Query(
+		"SELECT file_id, file_oid, external_id, backend, name, type, size, sha512, updated_at FROM files WHERE backend != $1 AND NAME LIKE $2 || '%' AND deleted_at IS NULL",
+		r.defaultBackend.Name(), prefix,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var res []BackendFile
+	for rows.Next() {
+		var info backend.FileInfo
+		var backendName string
+		var externalID *string
```

Comment:
does `externalID` needs to be `*string` or could it be just `string`? not seeing the difference for other cases here.